Add index link to detail tab (Issue 981 & 323)

juzen2003 opened the pull request:
- Fixes #981. Fixes #323 .
- This needs to work with https://github.com/SETI/pds-webtools/pull/23
- Known problems & pending items:
1. We need shelves/links and shelves/info for files like metadata/CORSS_8xxx/CORSS_8001/CORSS_8001_profile_index.lbl or metadata/EBROCC_xxxx/EBROCC_0001/ES1_profile_index.lbl. Because their internal_link_info is returning an empty tuple which will result in missing .lbl (only .tab exists) in the return of opus_products. It will also cause error in import when calling opus_products. And the missing info folder will raise error in import when calling file.size_bytes. For now import_for_tests will not import the following volsets due to the error mentioned above:
    - EBROCC_xxxx
    - CORSS_8xxx
    - COUVIS_8xxx
    - COVIMS_8xxx
2. The following opus types need proper tooltips in internal_def_product_types.json:
    - hstfiles_index
    - raw_image_index
    - profile_index
    - supplemental_index
3. Add profiling to import
4. Fix OPUS tests

juzen2003 committed:
Add tooltips for index (rms_index) product type in internal_def_product_types.json.

juzen2003 committed:
Make sure rms_index product type is unchecked by default in cart download
pane.
- In settings.py, create a global variable UNCHECKED_PRODUCT_TYPES to store
  a list of opus_types that are unchecked by default in the cart download
  pane. (line 373-374)
- In views.py, don't not check the checkbox when the product type is in
  settings.UNCHECKED_PRODUCT_TYPES. (line 124)

juzen2003 committed:
Look up index row selections when doing import, add an extra import parameter
to determine whether to use row files under shelves/index or not.
- In main_opus_import.py, add parameter --import-use-row-files to check
  if row files under shelves/index should be used. (line 301-305)
- In do_import.py, add the logic to look up index row selections when
  doing import. (line 1696-1728)

juzen2003 committed:
Construct the correct row entry url in detail tab.
- In views.py, update the logic to construct the correct row selection
  for row entry url in detail tab. (line 512-518)

juzen2003 committed:
Add an extra column to obs_files table to accommodate the extra element
"default_checked" in the keys of opus_products. This will be used to
determine whether an opus_product is checked by default in OPUS cart tab.
- In obs_files.json, add "default_checked" field (0 or 1) to obs_files
  table schema. (line 91-95)
- In models.py, add a field "default_checked" to ObsFiles model. (line 1356)
- In cart/views.py, if returned type has default_checked value equal to
  1, we will select the checkbox in the cart download pane UI. (line 124)
- In cart/views.py, in _get_download_info, query "default_checked", and
  pass it to product_dict_entry to be appended by cur_product_list.
  (line 784, 818)
- In file_utils, in get_pds_products, modify the SQL query to get "default_checked"
  as well, and modify the corresponding tuple constructure. (line79 ,
  114, 117, 120)
- In do_import.py, if "default_checked" is True, store 1 in database, else
  store 0. (line 1677-1679, 1762)

juzen2003 committed:
Remove redundant query for obsfile.default_checked in results/view.py.

juzen2003 committed:
Add placeholder tooltips for hstfiles_index, raw_image_index, profile_index,
and supplemental_index in internal_def_product_types.json, need to update
them with proper descriptions later.

rfrenchseti commented on opus/application/apps/ui/views.py:
```diff
@@ -505,9 +505,17 @@ def api_init_detail_page(request, **kwargs):
         for product_type in products[version]:
             file_list = products[version][product_type]
             product_info = {}
+            if (selection is None and (product_type[3].find('Browse Image') or
```
If there's room on the line, put the AND clauses on different lines:

```
if (X and
    (Y or Z)):
```

If not, then:

```
if (X and
    (Y or
     Z)):
```

rfrenchseti commented on opus/import/do_import.py:
```diff
@@ -1674,7 +1674,9 @@ def get_pdsfile_rows_for_filespec(filespec, obs_general_id, opus_id, volume_id,
     for product_type in products:
-        (category, sort_order_num, short_name, full_name) = product_type
+        (category, sort_order_num, short_name,
+            full_name, default_checked) = product_type
```
Fix indent

rfrenchseti commented on opus/application/apps/ui/views.py:
```diff
@@ -505,9 +505,17 @@ def api_init_detail_page(request, **kwargs):
             product_info = {}
+            if (selection is None and (product_type[3].find('Browse Image') or
+                product_type[3].find('Browse Diagram'))):
+                for fn in file_list:
```
This seems unnecessarily complicated. Since you always get the first filename no matter what, can't you just do:

```
if len(file_list) > 0:
    selection = file_list[0].split('/')[-1].split('.')[0]
```

rfrenchseti commented on opus/import/do_import.py:
```diff
@@ -1693,13 +1695,39 @@ def get_pdsfile_rows_for_filespec(filespec, obs_general_id, opus_id, volume_id,
         for sublist in list_of_sublists:
+            if (impglobals.ARGUMENTS.import_use_row_files and
+                skip_current_product_type):
+                break;
```
Remove ;

rfrenchseti commented on opus/import/do_import.py:
```diff
@@ -1693,13 +1695,39 @@ def get_pdsfile_rows_for_filespec(filespec, obs_general_id, opus_id, volume_id,
+                    ('_summary.tab' in logical_path or
+                     '_index.tab' in logical_path or
+                     '_hstfiles.tab' in logical_path)):
+
```
Remove blank line

rfrenchseti commented on opus/import/do_import.py:
```diff
@@ -1693,13 +1695,39 @@ def get_pdsfile_rows_for_filespec(filespec, obs_general_id, opus_id, volume_id,
+                     '_hstfiles.tab' in logical_path)):
+
+                    # Check if corresponding files exist in shelves/index
+                    if os.path.exists(file.indexshelf_abspath):
```
We shouldn't be looking deeply into PdsFile internals here. There should be a PdsFile function called something like "index_shelve_exists"

rfrenchseti commented on opus/import/do_import.py:
```diff
@@ -1693,13 +1695,39 @@ def get_pdsfile_rows_for_filespec(filespec, obs_general_id, opus_id, volume_id,
                 logical_path = file.logical_path
+
+                if (impglobals.ARGUMENTS.import_use_row_files and
+                    ('_summary.tab' in logical_path or
```
It really seems like there should be a better way to do this rather than hard-coding index filename suffixes here but I'm not sure what it would be.

rfrenchseti commented on opus/import/do_import.py:
```diff
@@ -1693,13 +1695,39 @@ def get_pdsfile_rows_for_filespec(filespec, obs_general_id, opus_id, volume_id,
+                        selection = basename.split('.')[0]
+                        try:
+                            file.find_selected_row_key(selection, '=')
+                        except: # can't find the row, we skip this producut_type
```
It's best to include the particular exception we're looking for rather than catching all of them, in case there's some other problem we end up hiding.

rfrenchseti replied:
Also misspelled "product_type"

rfrenchseti commented on opus/import/do_import.py:
```diff
@@ -1693,13 +1695,39 @@ def get_pdsfile_rows_for_filespec(filespec, obs_general_id, opus_id, volume_id,
+                            skip_current_product_type = True
+                            break
+                    else:
+                        impglobals.LOGGER.log('warning',
```
Take a look at nonrepeating_warning in import_util. This makes sure you only see a particular warning once.

rfrenchseti commented on opus/import/main_opus_import.py:
```diff
@@ -298,6 +298,11 @@
+# Argument about whether to check index row files or not
+parser.add_argument(
+    '--import-use-row-files', action='store_true', default=False,
+    help='Import with checking row files under shelves/index'
```
Let's make the help a little more user-friendly. Something like "Use metadata row files to determine whether index and summary files should be included in the files table".

rfrenchseti commented on opus/import/main_opus_import.py:
```diff
@@ -319,6 +324,8 @@
     impglobals.ARGUMENTS.copy_import_to_permanent_tables = True
     impglobals.ARGUMENTS.drop_new_import_tables = True
+    # Comment/Uncomment the line below to determine if row files are used
+    impglobals.ARGUMENTS.import_use_row_files = True
```
This shouldn't be here...it eliminates the benefit of having the option in the first place

rfrenchseti commented on opus/import/table_schemas/internal_def_product_types.json:
```diff
@@ -44,6 +44,31 @@
+    {
+        "pi_dict_context": "OPUS_PRODUCT_TYPE",
+        "pi_dict_name": "rms_index",
+        "definition": "Text files created by the Ring-Moon Systems Node that summarize or augment some metadata for all observations in a particular volume. Associated labels (*.lbl) describe the contents of the text files."
```
I'll look at these another time

rfrenchseti replied:
This is fine.

juzen2003 committed:
Update Django tests and add modifications based on comments.
- In views.py, update logic to obtain row selection. (line 516)
- In main_opus_import.py, add --profile to do performance profiling
  (line 310-312, 412-415, 523-533).
- In import_for_tests.sh, add parameter to check row files in each import
  command. (line 15-26)
- In do_import.py, update the logic to determine if we should import
  a product based on row files. Also check the existence of shelf info
  and links to determine if we should import a file. (line 1712-1737)

juzen2003 committed:
Merge remote-tracking branch 'upstream/master' into 981_323_add_index_link_to_detail_tab

juzen2003 committed:
Merge remote-tracking branch 'upstream/master' into 981_323_add_index_link_to_detail_tab

juzen2003 committed:
Properly modify filespec by replacing .LBL with .TAB for COUVIS_8xxx in
populate_obs_general_preview_images in populate_obs_general.py.

juzen2003 committed:
Fix the url for index row entry.
- In views.py, in the url of index row entry, replace "opus.pds-rings" with
  "pds-rings".

juzen2003 committed:
Fix the step in get_pdsfile_rows_for_filespec to check only shelves/info
instead of both shelves/info & shelves/links before accessing _info.
- In do_import.py, remove the step of checking shelves/links before
  accessing _info. This will make sure we don't skip preview images, and
  all preview images are properly imported to database.

juzen2003 committed:
Modify tests, update URL to have start with opus.pds-rings in test results.

juzen2003 committed:
Fix the issue of multiple imports on one line in main_opus_import.py.

rfrenchseti commented on opus/application/apps/cart/views.py:
```diff
@@ -120,7 +120,8 @@ def api_view_cart(request):
     for name, details in info['product_cat_list']:
         for type in details:
-            if type['slug_name'] in not_selected_product_types:
+            if(type['slug_name'] in not_selected_product_types or
```
Add space after if (and update indentation of next line)

rfrenchseti commented on opus/application/apps/ui/views.py:
```diff
@@ -519,7 +526,8 @@ def api_init_detail_page(request, **kwargs):
                     tab_url = tab_url.replace('holdings', 'viewmaster')
-                    tab_url += '/'+opus_id.split('-')[-1]
+                    tab_url += '/'+selection
+                    tab_url = tab_url.replace('opus.pds-rings', 'pds-rings')
```
We shouldn't be hard-coding web addresses in the code. opus.pds-rings is defined in settings.py. You can add a new global variable there for where to find ViewMaster, then use those two variables to do the replace here.

rfrenchseti commented on opus/application/apps/ui/views.py:
```diff
@@ -505,9 +505,16 @@ def api_init_detail_page(request, **kwargs):
     new_products = OrderedDict()
     for version in products:
         new_products[version] = OrderedDict()
+        selection = None
```
I understand why you're doing this, but it's the wrong solution. What if an observation doesn't even have a Browse Image? The right thing to do is retrieve the filespec from the obs_pds table (see how to get things from ObsGeneral above). Then derive the selection suffix from the filespec.

rfrenchseti commented on opus/import/do_import.py:
```diff
@@ -1682,7 +1682,9 @@ def get_pdsfile_rows_for_filespec(filespec, obs_general_id, opus_id, volume_id,
-        (category, sort_order_num, short_name, full_name) = product_type
+        (category, sort_order_num, short_name,
+         full_name, default_checked) = product_type
+        default_checked = 1 if default_checked else 0
```
default_checked is already a 1 or 0. This line isn't needed.

rfrenchseti commented on opus/import/do_import.py:
```diff
@@ -1701,13 +1703,47 @@ def get_pdsfile_rows_for_filespec(filespec, obs_general_id, opus_id, volume_id,
+        skip_current_product_type = False
+
         for sublist in list_of_sublists:
+            if (impglobals.ARGUMENTS.import_use_row_files and
```
When we first started on this project, we didn't have index files, so this argument made sense. Now we have index files for everything, so it's sense should be inverted: import_dont_use_row_files. The default should be to use them now.

rfrenchseti commented on opus/import/do_import.py:
```diff
@@ -1701,13 +1703,47 @@ def get_pdsfile_rows_for_filespec(filespec, obs_general_id, opus_id, volume_id,
                     version_name = 'Current'
                 logical_path = file.logical_path
+
+                # is_index: check if it's an index file by the presence of the
```
This is a confusing comment. We check if it's an index file by the presence of some other file, but I don't see us checking for that file is. Is that what file.is_index does? How about changing the comment to say something like for an index file we check to see if this observation is present, and if not we don't include the index file in the results.

rfrenchseti commented on opus/import/do_import.py:
```diff
@@ -1701,13 +1703,47 @@ def get_pdsfile_rows_for_filespec(filespec, obs_general_id, opus_id, volume_id,
+                    selection = basename.split('.')[0]
+                    try:
+                        file.find_selected_row_key(selection, '=')
+                    except OSError:
```
This is really horrible. Isn't there some better way to see if the row key is in the index that doesn't throw a generic  OSError? Shouldn't find_selected_row_key return None or something?

rfrenchseti commented on opus/import/do_import.py:
```diff
@@ -1701,13 +1703,47 @@ def get_pdsfile_rows_for_filespec(filespec, obs_general_id, opus_id, volume_id,
+                try:
+                    file.shelf_lookup('info')
+                except (IOError, OSError, KeyError, ValueError):
+                    continue
```
Should we report some kind of warning to the user here?

rfrenchseti commented on opus/import/import_for_tests.sh:
```diff
@@ -12,18 +12,18 @@ if [ "$yn" != "YES" ]; then
-python main_opus_import.py --do-all-import CORSS_8001 $1
-python main_opus_import.py --do-all-import COUVIS_8001 $1
-python main_opus_import.py --do-all-import COVIMS_8001 $1
+python main_opus_import.py --do-all-import --import-use-row-files COISS_2002,COISS_2008,COISS_2111 $1
```
With the requested change above to --import-dont-use-row-files these changes can go away.

rfrenchseti commented on opus/import/main_opus_import.py:
```diff
@@ -503,6 +521,17 @@ def _new_warning_handler(message, category, filename, lineno, file, line):
+        ps.print_stats()
+        ps.print_callers()
+        impglobals.LOGGER.info('Profile results:\n%s', s.getvalue())
+        # with open('cprofile_results.txt', 'w+') as f:
```
Remove comments

rfrenchseti commented on opus/import/table_schemas/internal_def_product_types.json:
```diff
@@ -44,6 +44,31 @@
+    {
+        "pi_dict_context": "OPUS_PRODUCT_TYPE",
+        "pi_dict_name": "hstfiles_index",
+        "definition": "Text files ([volume]_hstfiles.tab) that list the associations between data files in a particular HST volume. Associated labels (*.lbl) describe the contents of the text files."
```
Text files ([volume]_hstfiles.tab) that list the FITS files in the STScI archive for this data set, along with the associated PDS files in this archive. Associated labels (*.lbl) describe the contents of the text files.

rfrenchseti commented on opus/import/table_schemas/internal_def_product_types.json:
```diff
@@ -44,6 +44,31 @@
+    {
+        "pi_dict_context": "OPUS_PRODUCT_TYPE",
+        "pi_dict_name": "raw_image_index",
+        "definition": "Text files ([volume]_raw_image_index.tab) that list the values of various raw image data in a particular volume. Associated labels (*.lbl) describe the contents of the text files."
```
Text files ([volume]_raw_image_index.tab) that tabulate information about the raw image for each observation in a particular volume. Associated labels (*.lbl) describe the contents of the text files.

rfrenchseti commented on opus/import/table_schemas/internal_def_product_types.json:
```diff
@@ -44,6 +44,31 @@
+    {
+        "pi_dict_context": "OPUS_PRODUCT_TYPE",
+        "pi_dict_name": "profile_index",
+        "definition": "Text files ([volume]_profile_index.tab) that [...placeholder]. Associated labels (*.lbl) describe the contents of the text files."
```
Text files ([volume]_profile_index.tab) that tabulate metadata from the data provider for all profiles in a set of occultations. Associated labels (*.lbl) describe the contents of the text files.

rfrenchseti commented on opus/import/table_schemas/internal_def_product_types.json:
```diff
@@ -44,6 +44,31 @@
+    {
+        "pi_dict_context": "OPUS_PRODUCT_TYPE",
+        "pi_dict_name": "supplemental_index",
+        "definition": "Text files ([volume]_supplemental_index.tab) that [...placeholder]. Associated labels (*.lbl) describe the contents of the text files."
```
Text files ([volume]_supplemental_index.tab) created by the Ring-Moon Systems Node that augment metadata for all observations in a particular volume. Associated labels (*.lbl) describe the contents of the text files.

juzen2003 committed:
Pass in fix_case=True to every from_filespec call in import.

juzen2003 committed:
Add variable VIEWMASTER_ROOT_PATH in settings.py so that we can properly
get the product path served by viewmaster.
- In settings.py, add a new variable VIEWMASTER_ROOT_PATH. (lin 310-311)
- In ui/views.py, replace product path with the one served by viewmaster
  by utilizing two variables in settings.py. (line 530-531)

juzen2003 committed:
Get the row selection based on filespec in obs general table when initializing
detail tab. Also, rename and modify the import parameter from "--import-use-row-files"
to "--import-dont-use-row-files".
- In views.py, get the row selection based on filesepc obtained from obs
  general table. (line 477,513,514)
- In do_import.py (line 1711,1723,1724), main_opus_import.py (line 306-308),
  import_for_tests.sh (line 15-26), rename and modify the import parameter
  from "--import-use-row-files" to "--import-dont-use-row-files".

juzen2003 committed:
Add tooltips for index files (hstfiles_index, raw_image_index, profile_index,
supplemental_index). Modify error handling to handler key error when checking
whether a selection is an index file.
- In internal_def_product_types.json, update tooltips for index files.
  (line 55,60,65,70)
- In main_opus_import.py, remove commented out codes. (line 532)
- In do_import.py, update error handling to handle key error when calling
  find_selected_row_key to check if a selection is in an index file.
  (line 1730)
- In do_import.py, add warning message when the corresponding shelves/info
  files don't exist. (line 1747-1748)
- In do_import.py, update comments. (line 1721-1723)

juzen2003 committed:
Skip all diagrams of COUVIS_8xxx in the import.
- In do_import.py, skip diagrams product of COUVIS_8xxx in the import.
  (line 1689-1692)

rfrenchseti commented on opus/application/apps/ui/views.py:
```diff
@@ -505,9 +506,13 @@ def api_init_detail_page(request, **kwargs):
         for product_type in products[version]:
             file_list = products[version][product_type]
             product_info = {}
+            if selection is None:
```
There's no reason for anything related to "selection" to be inside the loop. You just need to parse out the selection right after assigning filespec and don't need to do anything else.

rfrenchseti commented on opus/import/do_import.py:
```diff
@@ -1701,13 +1709,51 @@ def get_pdsfile_rows_for_filespec(filespec, obs_general_id, opus_id, volume_id,
+                    file.shelf_lookup('info')
+                except (IOError, OSError, KeyError, ValueError):
+                    import_util.log_warning('Missing corresponding ' +
+                                            f'shelves/info for {file}')
```
{file} is a PdsFile, right? Probably should put something more printable here - maybe file.abspath?

juzen2003 committed:
Move the step of obtaining the selection for index tab url right after
the step of getting filespec from obs general table.
- In views.py, obtain the selection for index tab url right after getting
  the filespec from obs general table. (line 478)
- In do_import.py, modify the warning message to print out the abspath
  of the file. (line 1753)

juzen2003 committed:
Add a new import parameter --import-report-empty-products to report empty
products during import.
- In main_opus_import.py, add a new parameter --import-report-empty-products.
  (line 316-319)
- In do_import.py, when --import-report-empty-products is enabled, display
  warning message when there is an empty product. (line 1679-1682)

juzen2003 committed:
Update expected default_checked values for tests in test_cart_api.py.

juzen2003 committed:
Reorganize --import-dont-use-row-files and --import-report-empty-products
to proper location in main_opus_import.py.

juzen2003 committed:
If an selection is partially matched in an index file, we will report a
warning message and skip that product type in the import.

juzen2003 committed:
Update shelf_lookup('info') to handle only OSError when the related
shelves/info files don't exist and abort the import when other errors
happened in do_import.py.

juzen2003 committed:
Update warning messages when selection is partially matched in an index
file during import.

juzen2003 committed:
Update the import warning messages when an index file has no files in
shelves/index. Also update the gain mode id mult options in obs_mission_hubble.
- In do_import.py, update the warning messages to specify which shelf file
  is missing. (line 1754)
- In obs_mission_hubble.json, update the mult options by removing "."
  (line 428-430)
